### DISCUSSION TITLE
Using "initial" texture value in shader

### DISCUSSION THREAD

**Discussion by mattjennings** [excaliburjs/Excalibur#2415] [Q&A] [answered]
I'm looking to create an effect similar to this [flashbang effect from FPS games](https://www.youtube.com/watch?v=RXZHCc2Po9c&t=113s) (1:55 mark). When the effect happens, it goes bright and fades out with an image of what your screen looked like right before the effect. How might I grab that "initial" screen texture so that I can fade it out overtop of the current texture? Is it possible to persist that value within the shader, or would I have to pass a texture in as a uniform manually?

**Comment by eonarheim** [comment, marked as answer]
Hi @mattjennings this is a good question! 

I think you might be able to do this kind of effect with a `PostProcessor` with custom shader code to perform the effect composition. The postprocessors don't have a super convenient way of updating uniforms, but you can hack around it by updating it in the `getLayout()`. I'll add an issue to make it more convenient to update postprocessor shader uniforms.

```typescript
class FlashPostProcessor implements PostProcessor {
  private _shader: ex.ScreenShader;
  constructor() {  }

  initialize(_gl: WebGLRenderingContext): void {
    this._shader = new ex.ScreenShader(
      `#version 300 es
      precision mediump float;
      // game screen texture
      uniform sampler2D u_image;
      // the screen coordinates
      in vec2 v_texcoord;

      // time in seconds
      uniform float time;

      // output color
      out vec4 fragColor;
      void main() {
        vec4 screenColor =  texture(u_image, v_texcoord);
        // fade for 5 seconds
        float totalTime = 5.0;
        // between 0-1
        float effectAmount = clamp(time / totalTime, 0.0, 1.0);
        // blend the colors
        fragColor = (screenColor * effectAmount) + (vec4(1.0, 1.0, 1.0, 1.0) * (1.0 - effectAmount));
      }`);
  }

  private _elapsedMs = 0;
  update(elapsedMs: number) {
    this._elapsedMs += elapsedMs;
  }

  getShader(): ex.Shader {
    return this._shader.getShader();
  }

  getLayout(): ex.VertexLayout {
    // Technically you can upload uniforms in here
    this._shader.getShader().setUniformFloat('time', this._elapsedMs/1000);
    return this._shader.getLayout();
  }
}

var flash = new FlashPostProcessor();
game.graphicsContext.addPostProcessor(flash);
game.onPostUpdate = (_, elapsedMs) => {
  flash.update(elapsedMs);
}
```
![effect](https://user-images.githubusercontent.com/612071/179338844-ba94f518-6961-47ce-a715-4eb9e42b44f1.gif)

**Reply by eonarheim** [reply]
Issue https://github.com/excaliburjs/Excalibur/issues/2416

**Reply by mattjennings** [reply]
Ah sorry I should have been more specific on my question! I'll try to backup and clarify a bit. I'm trying to create a "cross fade" effect between 2 scenes, where I snap the last frame of the prior scene and fade it out once the new scene has loaded. The clip I posted was the best demonstration I could find as it was close to what I wanted _except_ the bright white fading, my apologies for being unclear there.

After I posted this discussion I managed to do an alternative approach using `engine.screenshot()`, drawing the image via `ex.Canvas`, and then fading out its opacity.

https://user-images.githubusercontent.com/8703090/179369414-c874d4e5-3da6-4c49-a59c-e4e58a600b83.mp4

```js
  // snippet of the actor used to create the effect

  onInitialize(engine: Engine): void {
    engine.screenshot().then((el) => {
      this.screenshot = el
    })

    this.el = new ScreenElement({
      x: 0,
      y: 0,
      z: this.z,
      width: engine.canvasWidth,
      height: engine.canvasHeight,
    })

    this.el.graphics.use(
      new ex.Canvas({
        width: engine.canvasWidth,
        height: engine.canvasHeight,
        draw: (ctx) => {
          if (this.screenshot) {
            ctx.drawImage(this.screenshot, 0, 0)
          }
        },
      })
    )
    this.addChild(this.el)
  }

  onPreUpdate() {
       // simple fade out
       this.progress += 0.1
       this.el.graphics.opacity = 1 - this.progress
   }
```

This totally works, but the screenshot is not always _quite_ what the canvas was rendering. i.e if you look closely, there seems to be some antialiasing that appears on the character once the screenshot is drawn (the character is likely on a half-pixel here, so perhaps this is a separate issue).

https://user-images.githubusercontent.com/8703090/179369660-f0707aa8-0520-4bb1-a898-ffc5c90f1bdc.mp4

Regardless, if I could somehow recreate this effect in a shader I think that would be ideal. But I'm not sure how to actually get that initial frame as a texture, as the only texture the shader receives is one of the current canvas.

**Reply by mattjennings** [reply]
scratch the part about screenshots having antialiasing - i was able to resolve it by capturing a hi-res screenshot with `engine.screenshot(true)` and drawing it to scale like so:

```js
     new ex.Canvas({
        quality: window.devicePixelRatio,
        width: engine.canvasWidth,
        height: engine.canvasHeight,
        draw: (ctx) => {
          if (this.screenshot) {
            ctx.drawImage(
              this.screenshot,
              0,
              0,
              engine.canvasWidth,
              engine.canvasHeight,
              0,
              0,
              engine.canvasWidth / window.devicePixelRatio,
              engine.canvasHeight / window.devicePixelRatio
            )
          }
        },
      })
```

so that actually more-or-less solves my situation, but if the effect is possible via shader that would be sweet too.

**Reply by eonarheim** [reply]
@mattjennings That canvas workaround is slick! The stuff you're doing for the scene transition is so cool, this is something that we've wanted to be built in to scene transitions in some way (or at least a documented pattern) for Excalibur. I agree a shader would be the most idea (especially from a performance lens), I've noted that we want a way to upload arbitrary images as textures in our shader lifecycle issue https://github.com/excaliburjs/Excalibur/issues/2416#issuecomment-1186283414

I'll look into the antialiasing on the screenshot, it might be a bug with how it it grabs the current screen image from the buffer? Perhaps AA settings are also wrong down the stack. Bug: https://github.com/excaliburjs/Excalibur/issues/2421

**Reply by eonarheim** [reply]
Do you have a link to the code reproduction I could include in the bug?

**Reply by mattjennings** [reply]
Thanks!! I'm hoping soon to have an update in the Merlyn discussion thread about these transitions. I'm curious what you'll think of the API and how/if it could work as a first-class feature.

I'll work on a repro for the screenshot anti-aliasing. I actually wonder if it's not a bug though - as the resolution of my game was 320x180, so if my character was on a half-pixel and a non-hi-res screenshot was taken, the antialiasing probably is expected?
